In [146]:
import numpy as np
import pandas as pd

# Dealing with the Marketing Information

### Reading database containing all the information

In [147]:
rd_station = pd.read_csv("../data/lead_complete_rdstation.csv", sep=",")

C:\Users\anafl\AppData\Local\Temp\ipykernel_50876\1553664775.py:1: DtypeWarning: Columns (3,4,5,6,7,8,11,12,15,17,24,32,33,35,36,37,40,41,42,43,45,47,48,50,52,53,55,56,57,58,60,63,64,67,69,70,72,73,74,75,76,77,78,80,81,83,85,86,87,88,89,91,92,93,94,95,96,97,100,101,102,103,104,106,108,110,111,112,113,114,115,116,117,118,120,122,123,124,125,126,127,130,131,132,133,134,136,137,138,139,140,141,142,143,144,145,147,148,149,151,152,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,170,172,173,175,176,177,180,181,183,184,186,189,191,192,193,195,196,197,199,200,201,202,203,204,206,207,208,209,210,213,214,215,216,217,218,219,220,222,223,224,225,226,227,229,230,232,234,235,237,239,241,242,244,247,248,250,251,252,253,254,255,256,257,258,260,262,264,265,266,267,268,269,270,271,272,273,275,276,277,278,279,280,281,282,283,284,285,286,287,288,289,290,291,292,294,295,296,297,298,300,301,305,306,307,308,309,310,311,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,33

In [148]:
rd_station.columns = rd_station.columns.str.lower().str.replace(' ', '_')

In [149]:
rd_station.columns

Index(['email', 'nome', 'telefone', 'celular', 'facebook', 'twitter',
       'linkedin', 'website', 'cargo', 'empresa',
       ...
       'você_tem_interesse_no_curso_de_pós-graduação_em_engenharia_da_manutenção?',
       'você_tem_interesse_no_curso_engenharia_e_gestão_de_ativos?',
       'você_tem_interesse_no_pós-mba_gestão_legal?',
       'você_vai_orientar_um_grupo_de_potenciais_negociadores,_profissionais_de_sua_empresa,_sobre_como_trocar_concessões_(5ª_etapa)._que_recomendações_você_lhe_faria?',
       'você_é_aluno(a)?', 'você_é_convidado_indumak?',
       'você_é_nosso_aluno_regular?', 'você_é_professor(a)_da_sustentare?',
       'área_de_atuação_da_empresa', 'área_de_atuação_de_interesse'],
      dtype='object', length=346)

## Understanding the dataset

In [71]:
rd_station.sample()

email                        nome telefone  \
40926  coralmedicinadotrabalho@gmail.com  Coral Medicina do Trabalho      NaN   

      celular facebook twitter linkedin website cargo empresa  ...  \
40926     NaN      NaN     NaN      NaN     NaN   NaN     NaN  ...   

       você_tem_interesse_no_curso_de_pós-graduação_em_engenharia_da_manutenção?  \
40926                                                NaN                           

      você_tem_interesse_no_curso_engenharia_e_gestão_de_ativos?  \
40926                                                NaN           

      você_tem_interesse_no_pós-mba_gestão_legal?  \
40926                                         NaN   

      você_vai_orientar_um_grupo_de_potenciais_negociadores,_profissionais_de_sua_empresa,_sobre_como_trocar_concessões_(5ª_etapa)._que_recomendações_você_lhe_faria?  \
40926                                                NaN                                                                                                                

      você_é_aluno(a)? você_é_convidado_indumak? você_é_nosso_aluno_regular?  \
40926              NaN                       NaN                         NaN   

      você_é_professor(a)_da_sustentare?  área_de_atuação_da_empresa  \
40926                                NaN                         NaN   

      área_de_atuação_de_interesse  
40926                          NaN  

[1 rows x 346 columns]

In [150]:
rd_station.shape

(41722, 346)

In [151]:
rd_station.isna().sum()

email                                     0
nome                                    698
telefone                              33416
celular                               28981
facebook                              41527
                                      ...  
você_é_convidado_indumak?             41699
você_é_nosso_aluno_regular?           41579
você_é_professor(a)_da_sustentare?    41699
área_de_atuação_da_empresa            41557
área_de_atuação_de_interesse          41722
Length: 346, dtype: int64

In [152]:
rd_station.dtypes

email                                  object
nome                                   object
telefone                               object
celular                                object
facebook                               object
                                       ...   
você_é_convidado_indumak?              object
você_é_nosso_aluno_regular?            object
você_é_professor(a)_da_sustentare?     object
área_de_atuação_da_empresa             object
área_de_atuação_de_interesse          float64
Length: 346, dtype: object

## Dropping the columns that won't need

In [153]:
print(rd_station.columns)

Index(['email', 'nome', 'telefone', 'celular', 'facebook', 'twitter',
       'linkedin', 'website', 'cargo', 'empresa',
       ...
       'você_tem_interesse_no_curso_de_pós-graduação_em_engenharia_da_manutenção?',
       'você_tem_interesse_no_curso_engenharia_e_gestão_de_ativos?',
       'você_tem_interesse_no_pós-mba_gestão_legal?',
       'você_vai_orientar_um_grupo_de_potenciais_negociadores,_profissionais_de_sua_empresa,_sobre_como_trocar_concessões_(5ª_etapa)._que_recomendações_você_lhe_faria?',
       'você_é_aluno(a)?', 'você_é_convidado_indumak?',
       'você_é_nosso_aluno_regular?', 'você_é_professor(a)_da_sustentare?',
       'área_de_atuação_da_empresa', 'área_de_atuação_de_interesse'],
      dtype='object', length=346)


In [154]:
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(rd_station['cidade'].value_counts())

cidade
Joinville                    2796
Jaraguá do Sul                201
São Paulo                     185
São Bento do Sul              130
Itajaí                         96
Curitiba                       52
Blumenau                       51
Rio de Janeiro                 49
Florianópolis                  41
Belo Horizonte                 30
São Francisco do Sul           28
Fortaleza                      25
Brasília                       25
Sete Lagoas                    21
Navegantes                     20
Vila Velha                     19
Franca                         18
Schroeder                      18
Salvador                       16
Itapoá                         15
Araquari                       15
Goiânia                        14
Campinas                       14
Guaramirim                     14
Recife                         14
Tijucas                        13
Jundiaí                        12
Vitória                        11
Balneário Piçarras             11
Teresin

In [155]:
#rd_station.notnull().sum().sort_values(ascending=False).head(70)
# To be able to look at all the lines
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(rd_station.notnull().sum().sort_values(ascending=False).head(70))

email                                                                                            41722
estágio_no_funil                                                                                 41722
total_de_conversões                                                                              41722
url_pública                                                                                      41722
status_para_comunicação_por_email                                                                41722
lead_scoring_-_interesse                                                                         41722
lead_scoring_-_perfil                                                                            41722
base_legal_para_comunicação                                                                      41660
eventos_(últimos_100)                                                                            41340
origem_da_última_conversão                                               

### If wanted to created a subset with the most responds field

In [156]:
top_50_filled = rd_station.notnull().sum().sort_values(ascending=False).head(50).index
rd_station_columns_responds_most = rd_station[top_50_filled]

## Creating a subset with the information on the lead <> Profile

In [157]:
rd_station_profile = rd_station[[
    'email', 'nome', 'lead_scoring_-_perfil', 'url_pública', 'estágio_no_funil', 'total_de_conversões',
    'lead_scoring_-_interesse', 'status_para_comunicação_por_email',
    'data_da_primeira_conversão', 'eventos_(últimos_100)', 'origem_da_última_conversão',
    'data_da_última_conversão', 'origem_da_primeira_conversão', 'curso_de_interesse:',
    'empresa', 'cargo.1', 'tags', 'data_da_última_oportunidade', 'cargo', 'estado',
    'valor_total_da_oportunidade_no_crm_(última_atualização)',
    'qualificação_da_oportunidade_no_crm_(última_atualização)',
    'etapa_do_funil_de_vendas_no_crm_(última_atualização)',
    'nome_do_responsável_pela_oportunidade_no_crm_(última_atualização)',
    'origem_da_oportunidade_no_crm_(última_atualização)',
    'cidade', 'possui_interesse_em_realizar_uma_pós-graduação_ou_mba?',
    'qual_é_a_sua_área_de_atuação?', 'selecione_o_curso_desejado:.1',
    'qual_a_sua_área_de_atuação?', 'data_de_nascimento',
    'desejo_receber_o_sustentare_news', 'qual_sua_área_de_interesse?',
    'como_você_conheceu_a_sustentare:', 'universidade', 'reside_em:',
    'cidade_de_residência:', 'interesse_em', 'qual_o_curso_de_interesse?'
]]

### Got different columns with representing the same thing, compiling all in one new columns

Represents City: 'Cidade', 'Reside em:' and 'Cidade de residência:'

In [158]:
# Using .loc because was receiving 'SettingWithCopyWarning'
rd_station_profile.loc[:, 'cidade_final'] = (
    rd_station_profile['cidade']
    .fillna(rd_station_profile['reside_em:'])
    .fillna(rd_station_profile['cidade_de_residência:'])
)


C:\Users\anafl\AppData\Local\Temp\ipykernel_50876\1712831611.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  rd_station_profile.loc[:, 'cidade_final'] = (


Represents Position at the company: 'Cargo.1' and 'Cargo'

In [159]:
rd_station_profile.loc[:, 'cargo_final'] = (
    rd_station_profile['cargo.1']
    .fillna(rd_station_profile['cargo'])
)

C:\Users\anafl\AppData\Local\Temp\ipykernel_50876\1555167880.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  rd_station_profile.loc[:, 'cargo_final'] = (


Represents area of work in the company: 'Qual a sua área de atuação?', 'Qual a sua área de atuação?'

In [160]:
rd_station_profile.loc[:, 'area_atuacao'] = (
    rd_station_profile['qual_é_a_sua_área_de_atuação?']
    .fillna(rd_station_profile['qual_a_sua_área_de_atuação?'])
)

C:\Users\anafl\AppData\Local\Temp\ipykernel_50876\814111575.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  rd_station_profile.loc[:, 'area_atuacao'] = (


Represents course interest: 'Curso de interesse:', 'Qual sua área de interesse?' and 'Interesse em'

In [161]:
rd_station_profile.loc[:, 'interesse_final'] = (
    rd_station_profile['curso_de_interesse:']
    .fillna(rd_station_profile['selecione_o_curso_desejado:.1'])
    .fillna(rd_station_profile['qual_sua_área_de_interesse?'])
    .fillna(rd_station_profile['interesse_em'])
    .fillna(rd_station_profile['qual_o_curso_de_interesse?'])
)

C:\Users\anafl\AppData\Local\Temp\ipykernel_50876\2982015989.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  rd_station_profile.loc[:, 'interesse_final'] = (


In [162]:
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(rd_station_profile.isna().sum())

email                                                                    0
nome                                                                   698
lead_scoring_-_perfil                                                    0
url_pública                                                              0
estágio_no_funil                                                         0
total_de_conversões                                                      0
lead_scoring_-_interesse                                                 0
status_para_comunicação_por_email                                        0
data_da_primeira_conversão                                             382
eventos_(últimos_100)                                                  382
origem_da_última_conversão                                             382
data_da_última_conversão                                               382
origem_da_primeira_conversão                                           382
curso_de_interesse:      

With this I was able to reduce 
- *Cidade* had 37078 null values now 36235
- *Cargo* had 27238 null values now 21926
- *Area Atuacao* had 39209 null values now 35800
- *Interesse Final* had 23536 null values now 22223

In [163]:
rd_station_profile = rd_station_profile.drop(columns=['cidade', 'reside_em:', 'cidade_de_residência:', 'cargo.1', 'cargo', 'qual_é_a_sua_área_de_atuação?', 'qual_a_sua_área_de_atuação?', 'curso_de_interesse:', 'qual_sua_área_de_interesse?', 'interesse_em', 'selecione_o_curso_desejado:.1'])

# Dealing with the CRM information

In [164]:
info_student_unimestre = pd.read_csv("../data/info_student_unimestre.csv")
first_cd_turma_student = pd.read_csv("../data/first_cd_turma_for_student_unimestre.csv")
dt_cadastro_student = pd.read_csv("../data/dt_cadastro_student_unimestre.csv")
name_student = pd.read_csv("../data/name_student_unimestre.csv")

In [165]:
info_student_unimestre.isna().sum()

cd_pessoa           0
ds_email            0
ds_cidade        2248
dt_nascimento    1156
min_dt_log       8144
dtype: int64

In [166]:
info_student_unimestre.shape

(12056, 5)

Trying to get less null values on the min_dt_log

In [134]:
merged_df = pd.merge(info_student_unimestre, dt_cadastro_student[['cd_pessoa', 'dt_cadastro']], on='cd_pessoa', how='left')
merged_df = pd.merge(merged_df, first_cd_turma_student[['cd_pessoa', 'min_date']], on='cd_pessoa', how='left')

info_student_unimestre['min_dt_log'] = merged_df['dt_cadastro'].fillna(merged_df['min_date']).fillna(info_student_unimestre['min_dt_log'])

In [135]:
info_student_unimestre.isna().sum()

cd_pessoa           0
ds_email            0
ds_cidade        2248
dt_nascimento    1156
min_dt_log        151
dtype: int64

Was able to reduce 
- *min_dt_log* had 9163 null values now 151

Creating a df to be only one row per ds_email (to avoid the duplicates values checked in the exploring file)

In [136]:
info_student_unimestre_unique = info_student_unimestre.groupby('ds_email').first().reset_index()

# Joining Marketing info with CRM

In [137]:
merge_left = rd_station_profile.merge(info_student_unimestre_unique, left_on='email', right_on='ds_email', how='left')

Creating a new column saying it's a student when a cd_pessoa is found

In [138]:
merge_left['aluno'] = (merge_left['cd_pessoa'].notnull()).astype(int)
merge_left['aluno'].value_counts()

aluno
0    37599
1     4123
Name: count, dtype: int64

## Got different columns with representing the same thing, compiling all in one new columns

Represents city of the person: 'ds_cidade', 'Cidade Final'

In [139]:
merge_left.loc[:, 'cidade'] = (
    merge_left['ds_cidade']
    .fillna(merge_left['cidade_final'])
)

Represents city of the person: 'ds_cidade', 'Cidade Final'

In [140]:
merge_left.loc[:, 'data_nascimento'] = (
    merge_left['dt_nascimento']
    .fillna(merge_left['data_de_nascimento'])
)

In [141]:
merge_left.sample()

email   nome lead_scoring_-_perfil  \
20052  dossantosdiogo88@gmail.com  Diogo                     c   

                                             url_pública estágio_no_funil  \
20052  http://app.rdstation.com.br/leads/public/4a3af...             Lead   

       total_de_conversões  lead_scoring_-_interesse  \
20052                    6                        60   

       status_para_comunicação_por_email data_da_primeira_conversão  \
20052                               True  2019-11-29 14:52:14 -0300   

                                   eventos_(últimos_100)  ... area_atuacao  \
20052  equipes-de-alta-performance-aula / infografico...  ...   Engenharia   

                        interesse_final ds_email cd_pessoa ds_cidade  \
20052  equipes-de-alta-performance-aula      NaN       NaN       NaN   

      dt_nascimento min_dt_log  aluno  cidade data_nascimento  
20052           NaN        NaN      0     NaN             NaN  

[1 rows x 40 columns]

With this I was able to reduce 
- *Cidade* had 35856 null values now 33986
- *Data Nascimento* had 38346 null values now 36117

Dropping columns with replicate values

In [142]:
merge_left = merge_left.drop(columns=['ds_cidade', 'cidade_final', 'dt_nascimento', 'data_de_nascimento'])

In [143]:
merge_left[merge_left.notna().all(axis=1)]

email              nome lead_scoring_-_perfil  \
5597    carolineklock@hotmail.com    Caroline Klock                     c   
16047  stella.bousfield@gmail.com  Stella Bousfield                     c   

                                             url_pública  estágio_no_funil  \
5597   http://app.rdstation.com.br/leads/public/8717b...  Lead Qualificado   
16047  http://app.rdstation.com.br/leads/public/1b271...           Cliente   

       total_de_conversões  lead_scoring_-_interesse  \
5597                    16                       120   
16047                   27                       170   

       status_para_comunicação_por_email data_da_primeira_conversão  \
5597                                True  2015-09-01 19:17:37 -0300   
16047                               True  2015-12-08 10:18:55 -0200   

                                   eventos_(últimos_100)  ...  \
5597   jornada-assessment-instrumentos-para-lideranca...  ...   
16047  webinar-storytelling / indicado-sarau-de-negoc...  ...   

               qual_o_curso_de_interesse?    cargo_final area_atuacao  \
5597   Educação empresarial / Remuneração  Outros Cargos           RH   
16047             Responsabilidade social  Outros Cargos    Marketing   

                   interesse_final                    ds_email    cd_pessoa  \
5597   formulario-de-pre-inscricao   carolineklock@hotmail.com     128609.0   
16047         webinar-storytelling  stella.bousfield@gmail.com  160000017.0   

                min_dt_log  aluno     cidade      data_nascimento  
5597   2017-08-21 00:00:00      1  Joinville  1985-09-21 00:00:00  
16047  2022-06-23 00:00:00      1  Joinville  1977-06-02 00:00:00  

[2 rows x 36 columns]

In [145]:
merge_left.isna().sum()

email                                                                    0
nome                                                                   698
lead_scoring_-_perfil                                                    0
url_pública                                                              0
estágio_no_funil                                                         0
total_de_conversões                                                      0
lead_scoring_-_interesse                                                 0
status_para_comunicação_por_email                                        0
data_da_primeira_conversão                                             382
eventos_(últimos_100)                                                  382
origem_da_última_conversão                                             382
data_da_última_conversão                                               382
origem_da_primeira_conversão                                           382
empresa                  

In [144]:
merge_left.to_csv("../data/merged_lead_student.csv")